src/python/pants/backend/scala/lint/scalafmt/rules.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+import textwrap
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.scala.lint.scala_fmt import ScalaFmtRequest
+from pants.backend.scala.lint.scalafmt.skip_field import SkipScalafmtField
+from pants.backend.scala.lint.scalafmt.subsystem import ScalafmtSubsystem
+from pants.backend.scala.target_types import ScalaSourceField
+from pants.core.goals.fmt import FmtResult
+from pants.core.goals.lint import LintRequest, LintResult, LintResults
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests
+from pants.engine.internals.selectors import Get, MultiGet
+from pants.engine.process import BashBinary, FallibleProcessResult, Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.jvm.jdk_rules import JdkSetup
+from pants.jvm.resolve.coursier_fetch import MaterializedClasspath, MaterializedClasspathRequest
+from pants.jvm.resolve.jvm_tool import JvmToolLockfileRequest, JvmToolLockfileSentinel
+from pants.testutil.rule_runner import logging
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class ScalafmtFieldSet(FieldSet):
+    required_fields = (ScalaSourceField,)
+
+    source: ScalaSourceField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipScalafmtField).value
+
+
+class ScalafmtRequest(ScalaFmtRequest, LintRequest):
```

Comment:
@stuhood / @chrisjrn: I don't like the naming here (`ScalafmtRequest` as concrete impl of a `ScalaFmtRequest`). Any thoughts on a better name?

Reply:
To be clear, is `ScalaFmtRequest` the concept of "generically format a Scala file", and `ScalafmtRequest` the concept of "use the tool `scalafmt`"? 

If so, I think `ScalaFmt` should be `ScalaFmtGoal`, and `Scalafmt` be `ScalafmtTool`, and just be explicit about both everywhere.

Reply:
(noting that my suggestion isn't one of consistency with the rest of the pants codebase, but attempting consistency here would be a [foolish one](https://en.wikipedia.org/wiki/Self-Reliance), imo)

Reply:
> To be clear, is `ScalaFmtRequest` the concept of "generically format a Scala file", and `ScalafmtRequest` the concept of "use the tool scalafmt"?

Correct. The naming convention for the language union is `LangFmtRequest` and the naming convention for the individual tools is `ToolRequest`, which breaks when the tool has "fmt" in its name. For Go, we called it `GoLangFmtRequest` so maybe `ScalaLangFmtRequest` for the language union?

Reply:
This sounds good to me!

Reply:
Yep, `Lang` seems fine.

Also, FWIW: the original reason for the `Lang` indirection and the extra per-language format union was that `Python` has a crazy number of formatters. For languages where we expect only one formatter to be enabled at once (...probably both `Java`/`Scala`/`Rust`/`Go`), the extra complexity of the indirection might not be necessary. Since these rules live in core, we can change their public APIs aggressively in future if we need to (unless we're expecting out-of-core formatting `@rules`).

Reply:
> For languages where we expect only one formatter to be enabled at once (...probably both `Java`/`Scala`/`Rust`/`Go`), the extra complexity of the indirection might not be necessary. 

Some users may have local/private linters that they want to hook in. There is a benefit to having a consistent experience in extending Pants across all languages for such use cases.

Reply:
Also `core/goals/fmt.py` is already pretty complex. +1 to still aggregating all formatters for a language together, even if it's only 1 formatter.

I agree here with calling it `ScalaLangFmtRequest`